cpp/src/cylon/table.cpp

```diff
@@ -1469,19 +1467,28 @@ Status Equals(const std::shared_ptr<cylon::Table> &a, const std::shared_ptr<cylo
 static Status RepartitionToMatchOtherTable(const std::shared_ptr<cylon::Table> &a,
                                            const std::shared_ptr<cylon::Table> &b,
                                            std::shared_ptr<cylon::Table> *b_out) {
-  int world_size = a->GetContext()->GetWorldSize();
   int64_t num_row = a->Rows();
 
   if (num_row == 0) {
     *b_out = a;
     return Status::OK();
   }
 
-  std::vector<int64_t> rows_per_partition;
-  RETURN_CYLON_STATUS_IF_FAILED(mpi::AllGather(a->GetContext(),
-                                               {num_row},
-                                               world_size,
-                                               rows_per_partition));
+  std::vector<int64_t> num_row_v = {num_row}, rows_per_partition;
+
+  std::shared_ptr<cylon::Column> num_row_col;
+  RETURN_CYLON_STATUS_IF_FAILED(Column::FromVector(num_row_v, num_row_col));
+
+  std::vector<std::shared_ptr<cylon::Column>> output;
+
+  RETURN_CYLON_STATUS_IF_FAILED(
+      a->GetContext()->GetCommunicator()->Allgather(num_row_col, &output));
+
+  for (auto col : output) {
+    auto temp = std::static_pointer_cast<arrow::Int64Scalar>(
+        col->data()->GetScalar(0).ValueOrDie());
+    rows_per_partition.push_back(temp->value);
+  }
```

Comment:
I think you should've used the `Scalar` all-gather here. 
```suggestion
  const auto& ctx = a->GetContext();
  auto num_row_v = Scalar::Make(arrow::MakeScalar(num_row));
  std::shared_ptr<Column> num_row_col;
  RETURN_CYLON_STATUS_IF_FAILED(ctx->GetCommunicator()->Allgather(num_row_v, &num_row_col));
  
  int64_t* vals = std::static_pointer_cast<arrow::Int64Array>(num_row_col->data())->raw_values();
  
  // TODO: it would have been nice, if you could use int64_t* vals, and int64_t num_row_col->length() instead of a vector
  // creating a vector here, copies data again 
  std::vector<int64_t> rows_per_partition(vals, vals + num_row_col->length()); 
```